app/maptools/recordingmaptool.cpp

```diff
@@ -568,19 +566,69 @@ void RecordingMapTool::releaseVertex( const QgsPoint &point )
     return;
   }
 
-  if ( mRecordedGeometry.type() == QgsWkbTypes::PolygonGeometry && mRecordedGeometry.constGet()->nCoordinates() == 2 )
-  {
-    setState( MapToolState::Record );
-  }
-  else if ( mRecordedGeometry.constGet()->nCoordinates() == 1 )
+  int vertexCount = mRecordedGeometry.constGet()->vertexCount( mActiveVertex.vertexId().part, mActiveVertex.vertexId().ring );
+
+  if ( mRecordedGeometry.type() == QgsWkbTypes::PolygonGeometry )
   {
-    updateVertex( mActiveVertex, point );
-    setState( MapToolState::Record );
+    QgsLineString *r;
+
+    if ( mActiveVertex.vertexId().ring == 0 )
+    {
+      r = qgsgeometry_cast<QgsLineString *>( qgsgeometry_cast<const QgsPolygon *>( mRecordedGeometry.constGet() )->exteriorRing() );
+    }
+    else
+    {
+      r = qgsgeometry_cast<QgsLineString *>( qgsgeometry_cast<const QgsPolygon *>( mRecordedGeometry.constGet() )->interiorRing( mActiveVertex.vertexId().ring ) );
+    }
+
+    if ( !r )
+    {
+      return;
+    }
+
+    if ( vertexCount == 2 )
+    {
+      r->close();
+      emit recordedGeometryChanged( mRecordedGeometry );
+
+      setState( MapToolState::Record );
+      setActiveVertex( Vertex() );
+      return;
+    }
+    else if ( vertexCount == 1 )
+    {
+      updateVertex( mActiveVertex, point );
+      setState( MapToolState::Record );
+      setActiveVertex( Vertex() );
+      return;
+    }
   }
-  else
+
+  updateVertex( mActiveVertex, point );
+
+  // if it is a first or last vertex of the line we go to the recording mode
+  if ( mRecordedGeometry.type() == QgsWkbTypes::LineGeometry )
   {
-    setState( MapToolState::View );
+    if ( mActiveVertex.type() != Vertex::Existing && mActiveVertex.vertexId().vertex == 0 )
+    {
+      setState( MapToolState::Record );
+      mActiveVertex.setCoordinates( QgsPoint() );
```

Comment:
I would say we should invalidate the active index altogether `setActiveVertex( Vertex() );`

Reply:
And also here we want to call `mNewVertexOrder(NewVertexOrder::Start)` to start digitizing from the beginning

Reply:
Hmm, not sure this is required. Recording works fine with valid acrive index, e.g. when handle is selected. Also when active vertex is set there is no need to set mNewVertexOrder.

Reply:
I believe active vertex should only be valid in `Grab` state and nowhere else. Just like you changed it 🙂